src/ice_spi.c

```diff
@@ -55,42 +60,29 @@ void ice_spi_init(void) {
     gpio_set_dir(ICE_SPI_SCK_PIN, GPIO_IN);
     gpio_set_dir(ICE_SPI_TX_PIN, GPIO_IN);
     gpio_set_dir(ICE_SPI_RX_PIN, GPIO_IN);
+
+    // Initialize SPI, but don't yet assign the pins SPI function so they stay in high impedance mode.
+    spi_init(spi1, 50 * 1000 * 1000);
+
+    // Setup DMA channel and interrupt handler
+    dma_tx = dma_claim_unused_channel(true);
+    dma_rx = dma_claim_unused_channel(true);
+
+    dma_channel_set_irq1_enabled(dma_rx, true);
+    irq_set_exclusive_handler(DMA_IRQ_1, spi_irq_handler);
+    irq_set_enabled(DMA_IRQ_1, true);
 }
 
 void ice_spi_init_cs_pin(uint8_t csn_pin) {
     gpio_init(csn_pin);
     ice_spi_chip_deselect(csn_pin);
 }
 
-static uint8_t transfer_byte(uint8_t tx) {
-    uint8_t rx;
-
-    for (uint8_t i = 0; i < 8; i++) {
-        // Update TX and immediately set negative edge.
-        gpio_put(ICE_SPI_SCK_PIN, false);
-        gpio_put(ICE_SPI_TX_PIN, tx >> 7);
-        tx <<= 1;
-
-        // stable for a while with clock low
-        sleep_us(1);
-
-        // Sample RX as we set positive edge.
-        rx <<= 1;
-        rx |= gpio_get(ICE_SPI_RX_PIN);
-        gpio_put(ICE_SPI_SCK_PIN, true);
-
-        // stable for a while with clock high
-        sleep_us(1);
-    }
-    return rx;
-}
-
 void ice_spi_chip_select(uint8_t csn_pin) {
     // Take control of the bus
-    gpio_put(ICE_SPI_SCK_PIN, false);
-    gpio_put(ICE_SPI_TX_PIN, true);
-    gpio_set_dir(ICE_SPI_SCK_PIN, GPIO_OUT);
-    gpio_set_dir(ICE_SPI_TX_PIN, GPIO_OUT);
+    gpio_set_function(ICE_SPI_RX_PIN, GPIO_FUNC_SPI);
+    gpio_set_function(ICE_SPI_SCK_PIN, GPIO_FUNC_SPI);
+    gpio_set_function(ICE_SPI_TX_PIN, GPIO_FUNC_SPI);
```

Comment:
I think I'd like to keep the gpio_put and gpio_set_dir in addition to this, in case the application code uses the SPI lines for something else, and resume to SPI later.

Reply:
I don't think it makes sense for the user to mess with the SPI IOs while an SPI transaction is in progress.  The `ice_spi_chip_deselect` puts them back to GPIO control so that the user can use them between transactions.

Potentially the deselect should always set them back to inputs in case the user has messed with them.

Reply:
For instance, the test FPGA firmware change the status of all I/Os for the purpose of checking for solder blobs.

This is a very silly use-case, but makes sense in context? Though, I agree with you!